README.md
# [![Gem Version](https://badge.fury.io/rb/oxr.svg)](https://badge.fury.io/rb/oxr) [![Build Status](https://travis-ci.org/jparker/oxr.svg?branch=master)](https://travis-ci.org/jparker/oxr)

# OXR

This gem provides a basic interface to the
[Open Exchange Rates](https://openexchangerates.org) API. At present, only the
API calls available to free plans have been implemented.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'oxr'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install oxr

## Usage

If you have not done so already, sign up for account on
[Open Exchange Rates](https://openexchangerates.org). Once you have an account,
go to Your Dashboard and locate your App ID.

Configure OXR with your App ID by calling `OXR.configure`:

```ruby
OXR.configure do |config|
  config.app_id = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
  # config.base = 'USD'
end
```

(If you are using OXR within a Rails application, you will probably want to put
this in an initializer.)

You can get current exchange rates using `OXR.get_rate`:

```ruby
OXR.get_rate 'GBP' # => 0.703087
OXR.get_rate 'JPY' # => 111.7062
```

You can also use the `OXR.[]` shortcut method.

```ruby
OXR['GBP'] # => 0.703087
```

`OXR.get_rate` accepts an optional keyword argument to retrieve historical
exchange rates for given dates. The provided date should be an object which
responds to `#strftime`.

```ruby
OXR.get_rate 'GBP', on: Date.new(2015, 6, 14) # => 0.642607
```

You perform more complex operations by using the lower-level API calls. These
methods return the raw JSON responses returned by Open Exchange Rates (parsed
using the [json](https://rubygems.org/gems/json) gem).

Get the latest exchange rates with `OXR.latest`.

```ruby
OXR.latest
```

This will return a JSON object with a structure similar to the following:

```ruby
{
 "disclaimer"=>"Usage subject to terms: https://openexchangerates.org/terms",
 "license"=>"https://openexchangerates.org/license",
 "timestamp"=>1512842416,
 "base"=>"USD",
 "rates"=>
  {
    "AED"=>3.673097,
    "AFN"=>68.693,
    "ALL"=>113.595865,
    # ...
    "ZAR"=>13.65526, 
    "ZMW"=>10.243477,
    "ZWL"=>322.355011
  }
}
```

Get historical exchange rates for specific dates with `OXR.historical`. This
method requires you to provide the date you wish to lookup. The date argument
should respond to `#strftime`.

```ruby
OXR.historical on: Date.new(2016, 3, 24)
```

This will return a JSON object with a structure similar to that returned by `OXR.latest`.

Get a list of available currencies with `OXR.currencies`.

```ruby
OXR.currencies
```

Get information about your account including your usage for the current period
with `OXR.usage`.

```ruby
OXR.usage
```

## Testing

Normally, any API call will send a request to Open Exchange Rates. Since your
plan allows a limited number of requests per month, you probably want to avoid
this when running in a test environment. You can stub the responses of specific
API calls by configuring the endpoint for specific calls to use a local file
instead of an HTTP request. Just provide a JSON file that reflects the payload
of an actual API call. You may provide this as a Pathname or a URI object, but
a plain old String will work as well. (You will find usable JSON files in
test/fixtures included with this gem.)

When you're done, you can call `OXR.reset_sources` to restore the default behavior.

Below is an example of configuring OXR within a test.

```ruby
class SomeTest < Minitest::Test
  def setup
    OXR.configure do |config|
      config.latest = 'test/fixtures/sample.json'
    end
  end

  def teardown
    OXR.reset_sources
  end
end
```

(You might consider doing this in your development environment as well.)

## Upgrading

The interface has changed between the 0.2.0 and 0.3.0 tags. It is no longer
necessary to instantiate an OXR object. The API calls are available as class
methods directly on the OXR module.

```ruby
# Before
oxr = OXR.new 'YOUR_APP_ID'
oxr['JPY']
oxr.usage

# Now
OXR.configure do |config|
  config.app_id = 'YOUR_APP_ID'
end
OXR['JPY']
OXR.usage
```

(You can still call `OXR.new`, but this behavior will generate a deprecation
warning.)

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run
`rake test` to run the tests. You can also run `bin/console` for an interactive
prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To
release a new version, update the version number in `version.rb`, and then run
`bundle exec rake release`, which will create a git tag for the version, push
git commits and tags, and push the `.gem` file to
[rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at
https://github.com/jparker/oxr.


## License

The gem is available as open source under the terms of the [MIT
License](http://opensource.org/licenses/MIT).


CHANGELOG.md
## OXR 0.6.1 (April 16, 2021) ##

* Accept endpoints as Pathnames or URIs as well as Strings
* Eliminated redundant URI processing for default endpoints

## OXR 0.6.0 (April 15, 2021) ##

*   Now works on Ruby 3.0
*   Gem now requires Ruby >= 2.5, only testing >= 2.6
*   Addressed Rubocop warnings
*   Code comments

## OXR 0.5.0 (December 16, 2017) ##

*   Added `base` option for performing queries with custom base currencies.

    *Techbrunch*


test/oxr_test.rb
# frozen_string_literal: true

require 'test_helper'

class OXRTest < Minitest::Test
  # rubocop:disable Metrics/AbcSize
  def test_missing_app_id
    OXR.configure { |config| config.app_id = nil }
    response = File.expand_path './fixtures/missing_app_id.json', __dir__
    stub_request(:get, %r{\Ahttps://openexchangerates.org/api/.*})
      .with(query: { 'app_id' => '' })
      .to_return status: 401, body: File.open(response)

    error = assert_raises(OXR::ApiError) { OXR['USD'] }
    assert_match(/No App ID provided/, error.description)
    assert_equal 'missing_app_id', error.response['message']
    assert_equal 401, error.response['status']
  end
  # rubocop:enable Metrics/AbcSize

  def test_invalid_app_id
    response = File.expand_path './fixtures/invalid_app_id.json', __dir__
    stub_request(:get, %r{\Ahttps://openexchangerates.org/api/.*})
      .to_return status: 401, body: File.open(response)

    error = assert_raises(OXR::ApiError) { OXR['USD'] }
    assert_match(/Invalid App ID/, error.description)
    assert_equal 'invalid_app_id', error.response['message']
    assert_equal 401, error.response['status']
  end

  def test_access_restricted
    response = File.expand_path './fixtures/access_restricted.json', __dir__
    stub_request(:get, %r{\Ahttps://openexchangerates.org/api/.*})
      .to_return status: 429, body: File.open(response)

    error = assert_raises(OXR::ApiError) { OXR['USD'] }
    assert_match(/Access restricted/, error.description)
    assert_equal 'access_restricted', error.response['message']
    assert_equal 429, error.response['status']
  end

  # rubocop:disable Metrics/AbcSize
  def test_invalid_base
    OXR.configure { |config| config.base = 'XXX' }
    response = File.expand_path './fixtures/invalid_base.json', __dir__
    stub_request(:get, %r{\Ahttps://openexchangerates.org/api/.*})
      .to_return status: 400, body: File.open(response)

    error = assert_raises(OXR::ApiError) { OXR['USD'] }
    assert_match(/Invalid `base` currency \[XXX\]/, error.description)
    assert_equal 'invalid_base', error.response['message']
    assert_equal 400, error.response['status']
  ensure
    OXR.configure { |config| config.base = nil }
  end
  # rubocop:enable Metrics/AbcSize
end


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'oxr'

require 'minitest/autorun'
require 'webmock/minitest'
require 'pry'


test/oxr/usage_test.rb
# frozen_string_literal: true

require 'test_helper'

module OXR
  class UsageTest < Minitest::Test
    def setup
      OXR.configure { |config| config.app_id = 'XXX' }
      @endpoint = 'https://openexchangerates.org/api/usage.json'
    end

    def test_usage
      response = File.expand_path '../fixtures/usage.json', __dir__
      stub_request(:get, @endpoint)
        .with(query: { 'app_id' => 'XXX' })
        .to_return status: 200, body: File.open(response)

      data = OXR.usage
      refute_nil data.dig('data', 'usage', 'requests')
      refute_nil data.dig('data', 'usage', 'requests_quota')
      refute_nil data.dig('data', 'usage', 'requests_remaining')
    end
  end
end


lib/oxr/configuration.rb
# frozen_string_literal: true

require 'uri'

module OXR
  ##
  # A container for OXR configuration options.
  #
  # Stores the API application ID and the base currency for which to request
  # exchange rates.
  #
  # Also stores the URI endpoints for fetching currencies, the latest exchange
  # ranges,  historical exchange rates, and current account usage statistics.
  #
  # By default, endpoints are generated dynamically, allowing them to
  # automatically adapte to changes to the app_id, but they may be set to a
  # fixed path, including a local file path. This is useful during testing when
  # you might want deterministic results and do not want to waste real API
  # requests.
  class Configuration
    ENDPOINT   = 'https://openexchangerates.org/api/'
    LATEST     = URI.join(ENDPOINT, 'latest.json').freeze
    HISTORICAL = URI.join(ENDPOINT, 'historical/').freeze
    USAGE      = URI.join(ENDPOINT, 'usage.json').freeze
    CURRENCIES = URI.join(ENDPOINT, 'currencies.json').freeze

    ##
    # Get and set the application ID that will be sent to the API server.
    attr_accessor :app_id

    ##
    # Get and set the base currency to be used when fetching exchange rates.
    attr_accessor :base

    ##
    # Set respective API endpoints. Use these if you want to sidestep the API
    # server, e.g., for testing. The endpoint may be provided as a URI,
    # Pathname, or String.
    #
    # Setting an endpoint to +nil+ will restore the default value.
    attr_writer :currencies, :historical, :latest, :usage

    def initialize
      reset_sources
    end

    ##
    # Returns the endpoint for listing known currencies.
    def currencies
      @currencies || append_query(CURRENCIES)
    end

    ##
    # Returns the endpoint for historical currency exchange rates on the given
    # date.
    #
    # Expects +date+ to respond #strftime.
    def historical(date)
      @historical || append_query(URI.join(HISTORICAL, "#{date.strftime('%F')}.json"), base: base)
    end

    ##
    # Returns the endpoint for the latest currency exchange rates.
    def latest
      @latest || append_query(LATEST, base: base)
    end

    ##
    # Returns the endpoint for fetch current API usage statistics.
    def usage
      @usage || append_query(USAGE)
    end

    ##
    # Resets all API endpoints back to their default (dynamic generation).
    def reset_sources
      @currencies = nil
      @historical = nil
      @latest     = nil
      @usage      = nil
    end

    private

    def append_query(uri, base: nil)
      uri.dup.tap do |u|
        u.query = "app_id=#{app_id}"
        u.query += "&base=#{base}" if base
      end.to_s
    end
  end
end


lib/oxr.rb
# frozen_string_literal: true

require 'oxr/version'
require 'oxr/configuration'

require 'date'
require 'json'
require 'open-uri'

##
# All use of OXR takes place through module methods.
#
# Before you can make API calls, you must configure OXR with your Open Exchange
# Rates App ID. Visit https://openexchangerates.org to sign up and generate one.
#
#   OXR.configure do |config|
#     config.app_id = 'XXXX'
#   end
#
# Optionally you may also specify the base currency for the conversion rates.
# The default base currency is USD.
#
#   OXR.configure do |config|
#     config.app_id = 'XXXX'
#     config.base = 'JPY'
#   end
#
# If you will be using OXR in a test environment, you may want to test the
# substitute local files for the API endpoints to provide deterministic results
# and avoid running over your Open Exchange Rates API request quota.
#
#   class ThisIsATest < Minitest::Test
#     def setup
#       OXR.configure do |config|
#         config.latest = 'test/fixtures/sample.json'
#       end
#     end
#
#     def teardown
#       OXR.reset_sources
#     end
#   end
#
# To quickly get an exchange rate from the base currency to a given currency,
# use `OXR.get_rate`.
#
#   OXR.get_rate 'GBP' # => 0.703087
#   OXR.get_rate 'JPY' # => 111.7062
#
# This method is aliased as `OXR.[]`.
#
#   OXR['GBP'] # => 0.703087
#
# Pass the `:on` keyword argument to get an historical exchange rate on a given
# date.
#
#   OXR.get_rate 'GBP', on: Date.new(2015, 6, 14) # => 0.642607
#
module OXR
  class Error < StandardError
  end

  ##
  # ApiError is raised when OXR encounters an error connecting to the Open
  # Exchange Rates API server.
  class ApiError < Error
    def message
      cause.message
    end

    def description
      response['description']
    end

    def response
      @response ||= JSON.parse cause.io.read
    end
  end

  class << self
    ##
    # DEPRECATED: TO BE REMOVED IN 0.7.
    #
    # OXR should no longer be instantiated explicitly.
    #
    # Call OXR.configure to specify the application ID and use the module
    # methods below to access the different API endpoints.
    def new(app_id)
      warn '[DEPRECATION] OXR.new is deprecated and will be removed from 0.7.' \
        " Use OXR class methods instead (from #{caller(1..1).first})."
      configure do |config|
        config.app_id = app_id
      end
      self
    end

    ##
    # Get the latest exchange rate for currency identified by +code+.
    #
    # If the optional +on+ keyword is provided, it instead returns the
    # historical exchange rate on the given date.
    def get_rate(code, on: nil)
      data = if on
               historical on: on
             else
               latest
             end
      data['rates'][code.to_s]
    end

    alias [] get_rate

    ##
    # Returns a Hash mapping currency codes to full currency names.
    def currencies
      call configuration.currencies
    end

    ##
    # Returns a Hash mapping currency codes to their historical exchange rate
    # on the date given by the +on+ keyword argument.
    #
    # The exchange rate is relative to the configured base currency.
    def historical(on:)
      call configuration.historical on
    end

    ##
    # Returns a Hash mapping currency codes to their latest exchange rate.
    #
    # The exchange rate is relative to the configured base currency.
    def latest
      call configuration.latest
    end

    ##
    # Returns a Hash containinng details about the Open Exchange Rates account
    # and current usage statistics.
    def usage
      call configuration.usage
    end

    ##
    # Resets API endpoints to their defaults.
    def reset_sources
      configure(&:reset_sources)
    end

    ##
    # Returns the OXR configuration struct.
    #
    # If given a block, the configuration will be yielded to the block.
    def configure
      yield configuration if block_given?
      configuration
    end

    ##
    # Returns the OXR configuration struct.
    def configuration
      @configuration ||= Configuration.new
    end

    private

    def call(endpoint)
      # This method makes it possible for a user to provide an endpoint path as
      # either a String, Pathname, or URI. We will massage it until we have
      # something that responds to #read.
      endpoint = URI.parse(endpoint.to_s) unless endpoint.respond_to?(:read)
      endpoint = File.open(endpoint.path) unless endpoint.respond_to?(:read)

      JSON.parse endpoint.read
    rescue OpenURI::HTTPError => e
      raise ApiError, e
    end
  end
end
